cmd/tailscale/cli: Improve integration with OS ssh tools

evilhamsterman opened the pull request:
This adds three flags to allow native ssh tools to communicate with Tailscale to check and validate the host keys of Tailscale managed ssh connections. All are flags on the `tailscale ssh` subcommand. The regular subcommand continues to work as normal

evilhamsterman commented:
This fixes #10185

DentonGentry reviewed (changes requested):
It is not at all clear to me that this is the right thing to do. I am marking it as requesting changes to make it clear that it is not to be submitted yet.

evilhamsterman commented:
If this helps I recorded a demo of what this does. Doesn't change any existing `tailscale ssh` functionality just allows it to better integration with native `ssh` tools https://asciinema.org/a/ZohQ5Jqb3oyxyrp4usNRgkRKO

evilhamsterman commented:
It could especially help people using the MacOS sandbox app, they still couldn't run the wrapper but they could get all the benefits with the built in ssh

bradfitz commented:
I think I'd prefer this under some other tailscale CLI command to export info in different formats, rather than further confuse people about what `tailscale ssh` does (it's already a top point of confusion...I don't want to make it worse)

evilhamsterman commented:
Interesting that the `tailscale ssh` command is a point of confusion. But if you don't want to overload it I don't think `tailscale export` makes much sense that seems to me like it would export the config of Tailscale or a JSON of your Tailnet.

I was looking at putting the config part that dumps the ssh_config snippet under `tailscale configure` like the Kubernetes configuration is. Not sure how well I could automate parsing and adding the config like you do with the kubeconf because there don't seem to be any decent Go ssh_config parsers that support the `Match` option, I think that's why many tools I've used that are similar just dump it and tell you to `>> ~/.ssh/config`

The `--known-hosts` option might make sense as another output option for status, similar to `--json`, something like `tailscale status --ssh-host-keys` maybe even optionally `tailscale status --ssh-host-keys <nodename>` to just dump a specific node's keys? I could see situations where that may be helpful for other automation.

The real magic though is the `--check` option, that just has to return 0 or 1 for `ssh` to use in the `Match` option. The trick is that it's easy after if you can use the `KnownHostsCommand` because then `ssh` can call Tailscale again for the known-hosts list, but if you have OpenSSH older than 8.4 you can't use that so it has to use a known-hosts file that's written to disk which the easiest way I could figure out to keep that updated is to write it every time `--check` is called. Hopefully the fall back is something you could remove eventually but since there's still support versions of MacOS and Windows that have OpenSSH 8.1 can't quite yet. With all that magic do still want it on `tailscale status`? Maybe `tailscale status --check-ssh`?

evilhamsterman commented:
Now that Thanksgiving is over I was able to get back to this. I rebased and moved the commands. The ssh snippet is moved to `tailscale configure sshconfig` similar to the kubeconfig option. It still only prints a snippet, though I think I may have an idea to automate that better. The known_hosts keys and the check are moved to `tailscale status --ssh-host-keys` and `tailscale status --check-ssh-host`

This would also fix #10348

I'm moving to a draft because I'd like to improve the automation with regards to the ssh config

evilhamsterman commented:
OK it now will append the config snippet directly into the `.ssh/config` or update it if it already exists and is out of date

evilhamsterman commented:
Any update on my status?

bradfitz commented on cmd/tailscale/cli/ssh.go:
```diff
@@ -211,6 +289,18 @@ var getSSHClientEnvVar = func() string {
 }
 
+// getSSHClientVersion returns the current version of the SSH client
+func getSSHClientVersion(ssh string) *semver.Version {
```
this is unused; remove?

bradfitz commented on cmd/tailscale/cli/ssh.go:
```diff
@@ -162,7 +174,41 @@ func writeKnownHosts(st *ipnstate.Status) (knownHostsFile string, err error) {
-func genKnownHosts(st *ipnstate.Status) []byte {
+// checkSSHClientVersion checks if the OpenSSH version supports the
+// KnownHostsCommand option and returns true if it does
+func checkSSHClientVersion(sshBin string) bool {
```
name this "sshSupportsKnownHostsCommand" ?

bradfitz commented on util/slicesx/slicesx.go:
```diff
@@ -77,3 +77,15 @@ func EqualSameNil[S ~[]E, E comparable](s1, s2 S) bool {
 	return true
 }
+
+// ReplaceBetween replaces the elements between start and end exclusive with the
+// given replacement. It returns a new slice with the replacement applied.
+func ReplaceBetween[S ~[]T, T any](s S, start, end int, replacement []T) S {
```
I'd rather not add this to slicesx. It seems pretty specialized and I don't foresee having other callers.

If we later did, then we could move it here.

bradfitz commented on cmd/tailscale/cli/ssh.go:
```diff
@@ -171,38 +217,70 @@ func genKnownHosts(st *ipnstate.Status) []byte {
 				continue
 			}
-			fmt.Fprintf(&buf, "%s %s\n", ps.DNSName, hostKey)
+			// Join all ps.TailscaleIPs as strings separated by commas.
+			ips := make([]string, len(ps.TailscaleIPs))
+			for i, ip := range ps.TailscaleIPs {
+				ips[i] = ip.String()
+			}
+			// Generate comma separated string of all possible names for the host.
+			n := strings.Join(append(ips, ps.DNSName, strings.TrimSuffix(ps.DNSName, "."), strings.Split(ps.DNSName, ".")[0]), ",")
+			fmt.Fprintf(&buf, "%s %s\n", n, hostKey)
```
what's this part of the change about? Is this related? Can this be sent out separately in its own PR?

evilhamsterman replied:
The known_hosts file format allows for a comma separated list of hostnames or patterns then a space then the key https://man.openbsd.org/sshd#SSH_KNOWN_HOSTS_FILE_FORMAT. This isn't necessary with your ssh wrapper because you take care of that yourself. But since people can ssh to a device using any of hostname, fqdn, or IP when using the match feature every thing has to be accounted for. This creates a host file that has all the possible names that someone could use to connect to a device and is necessary to the rest of the PR.

bradfitz commented on cmd/tailscale/cli/status.go:
```diff
@@ -56,18 +56,22 @@ https://github.com/tailscale/tailscale/blob/main/ipn/ipnstate/ipnstate.go
 		fs.BoolVar(&statusArgs.peers, "peers", true, "show status of peers")
 		fs.StringVar(&statusArgs.listen, "listen", "127.0.0.1:8384", "listen address for web mode; use port 0 for automatic")
 		fs.BoolVar(&statusArgs.browser, "browser", true, "Open a browser in web mode")
+		fs.StringVar(&statusArgs.checkSSHHost, "check-ssh-host", "", "Check if a host's SSH is managed by Tailscalse")
```
lowercase "check" for consistency. (looks like --browser is also wrong)

and clarify this changes the exit status of the command?

evilhamsterman commented:
Thank you for the review. This sat so long I ended up deleting my repo I'll update and either reopen or create a new one as soon as I can

evilhamsterman commented:
I've created a new fork and I'm working on updating it with the comments above then I'll open a new PR.
